File: Sources/APTradeInfrastructure/FinnhubEarningsRepository.swift
import Foundation
import APTradeApplication
import APTradeDomain

/// Live earnings source backed by Finnhub's `/calendar/earnings`. Error handling mirrors
/// `FinnhubNewsRepository`: 429 → `.rateLimited`, non-2xx → `.network`, decode/other →
/// mapped to `AppError`.
///
/// Responses are cached in-memory per (fromDay, toDay) for `ttl` (default 6h — earnings
/// dates don't move intraday) so the calendar screen, every detail screen, and the daily
/// notification check share one network call. `now` is the same injectable clock seam
/// `CachingMarketDataRepository` uses, defaulting to the real wall clock. Since this type
/// is a class (not an actor), the cache dictionary is guarded by an explicit lock rather
/// than actor isolation.
///
/// Cancellation translation (review note from Task 11): `URLSession` surfaces a cancelled
/// task as `URLError(.cancelled)`, not `CancellationError`. `FetchEarningsCalendarUseCase`'s
/// guard only rethrows `CancellationError` — every other failure degrades to `[]` — so this
/// adapter translates `.cancelled` to `CancellationError()` before it reaches the use case;
/// otherwise a cancelled SwiftUI task would silently degrade to "no earnings" instead of
/// propagating cancellation.
public final class FinnhubEarningsRepository: EarningsCalendarRepository, @unchecked Sendable {
    private struct CacheEntry {
        let events: [EarningsEvent]
        let at: Date
    }

    private let apiKey: String
    private let now: () -> Date
    private let ttl: TimeInterval
    private let base = "https://finnhub.io/api/v1"
    private let fetchData: (URL) async throws -> Data

    private let lock = NSLock()
    private var cache: [String: CacheEntry] = [:]

    public init(
        apiKey: String,
        session: URLSession = .shared,
        now: @escaping () -> Date = Date.init,
        ttl: TimeInterval = 6 * 60 * 60
    ) {
        self.apiKey = apiKey
        self.now = now
        self.ttl = ttl
        self.fetchData = { url in
            let (data, response) = try await session.data(from: url)
            if let http = response as? HTTPURLResponse {
                if http.statusCode == 429 { throw AppError.rateLimited }
                guard (200..<300).contains(http.statusCode) else { throw AppError.network }
            }
            return data
        }
    }

    /// Test seam: injects a fake `fetchData` closure instead of a real `URLSession`, so
    /// cache/TTL/cancellation behavior can be exercised without a network stub.
    init(
        apiKey: String,
        now: @escaping () -> Date = Date.init,
        ttl: TimeInterval = 6 * 60 * 60,
        fetchData: @escaping (URL) async throws -> Data
    ) {
        self.apiKey = apiKey
        self.now = now
        self.ttl = ttl
        self.fetchData = fetchData
    }

    public func earnings(fromDay: String, toDay: String) async throws -> [EarningsEvent] {
        let key = "\(fromDay)|\(toDay)"

        let cached = lock.withLock { cache[key] }
        if let cached, now().timeIntervalSince(cached.at) < ttl {
            return cached.events
        }

        guard var components = URLComponents(string: base + "/calendar/earnings") else {
            throw AppError.network
        }
        components.queryItems = [
            URLQueryItem(name: "from", value: fromDay),
            URLQueryItem(name: "to", value: toDay),
            URLQueryItem(name: "token", value: apiKey),
        ]
        guard let url = components.url else { throw AppError.network }

        let events: [EarningsEvent]
        do {
            let data = try await fetchData(url)
            events = try FinnhubEarningsMapper.events(from: data)
        } catch let urlError as URLError where urlError.code == .cancelled {
            throw CancellationError()
        } catch let error as AppError {
            throw error
        } catch {
            throw AppError.network
        }

        lock.withLock { cache[key] = CacheEntry(events: events, at: now()) }
        return events
    }
}

File: shared/src/commonMain/kotlin/com/aptrade/shared/infrastructure/FinnhubEarningsRepository.kt
package com.aptrade.shared.infrastructure

import com.aptrade.shared.application.EarningsCalendarRepository
import com.aptrade.shared.application.QuoteError
import com.aptrade.shared.domain.EarningsEvent
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.request.get
import io.ktor.client.statement.HttpResponse
import io.ktor.http.HttpStatusCode
import io.ktor.http.isSuccess
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

private const val DEFAULT_TTL_SECONDS = 6 * 60 * 60L // earnings dates don't move intraday

/**
 * Finnhub-backed [EarningsCalendarRepository]. Reuses the Yahoo adapter's per-platform
 * client ([defaultYahooHttpClient]) and [FinnhubNewsRepository]'s error-mapping idiom
 * (429 -> RateLimited, other non-2xx/malformed -> Network); CancellationException always
 * rethrows first.
 *
 * Responses are cached in-memory per (fromDay, toDay) for [ttlSeconds] (default 6h) so the
 * calendar screen, every detail screen, and the daily notification check share one network
 * call. `nowEpochSeconds` is the same injectable clock seam [FinnhubNewsRepository] uses,
 * defaulting to the real wall clock.
 */
class FinnhubEarningsRepository internal constructor(
    private val apiKey: String,
    private val client: HttpClient,
    private val nowEpochSeconds: () -> Long = { epochSecondsNow() },
    private val ttlSeconds: Long = DEFAULT_TTL_SECONDS,
) : EarningsCalendarRepository, AutoCloseable {

    // Production / Swift-harness entry point: builds the platform HTTP client via the
    // defaultYahooHttpClient expect/actual (Darwin on Apple, CIO on the JVM).
    constructor(apiKey: String) : this(apiKey, defaultYahooHttpClient())

    override fun close() { client.close() }

    private data class CacheEntry(val atEpochSeconds: Long, val events: List<EarningsEvent>)

    private val mutex = Mutex()
    private val cache = HashMap<Pair<String, String>, CacheEntry>()

    override suspend fun earnings(fromDay: String, toDay: String): List<EarningsEvent> {
        val key = fromDay to toDay
        mutex.withLock {
            cache[key]?.let { entry ->
                if (nowEpochSeconds() - entry.atEpochSeconds < ttlSeconds) return entry.events
            }
        }

        val response = try {
            client.get("https://finnhub.io/api/v1/calendar/earnings") {
                url {
                    parameters.append("from", fromDay)
                    parameters.append("to", toDay)
                    parameters.append("token", apiKey)
                }
            }
        } catch (e: CancellationException) {
            throw e
        } catch (e: Throwable) {
            throw QuoteError.Network(e.message ?: "request failed")
        }

        val events = FinnhubEarningsMapper.events(decodeCalendar(response))
        mutex.withLock { cache[key] = CacheEntry(nowEpochSeconds(), events) }
        return events
    }

    private suspend fun decodeCalendar(response: HttpResponse): FinnhubEarningsCalendarDTO {
        if (response.status == HttpStatusCode.TooManyRequests) throw QuoteError.RateLimited
        if (!response.status.isSuccess()) throw QuoteError.Network("HTTP ${response.status.value}")

        return try {
            finnhubJson.decodeFromString(
                FinnhubEarningsCalendarDTO.serializer(),
                response.body<String>(),
            )
        } catch (e: CancellationException) {
            throw e
        } catch (e: Throwable) {
            throw QuoteError.Network("malformed response")
        }
    }
}
